File: src/engines/calendar/jawa/jawaEngine.js
const DAY_MS = 86400000

const WULAN = [
  'Sura',
  'Sapar',
  'Mulud',
  'Bakdamulud',
  'Jumadilawal',
  'Jumadilakir',
  'Rejeb',
  'Ruwah',
  'Pasa',
  'Sawal',
  'Dulkangidah',
  'Besar',
]

const TAUN = [
  'Alip',
  'Ehe',
  'Jimawal',
  'Je',
  'Dal',
  'Be',
  'Wawu',
  'Jimakir',
]

const PAS = ['Pon', 'Wage', 'Kliwon', 'Legi', 'Pahing']

const MINGGU = [
  'Senen',
  'Selasa',
  'Rebo',
  'Kemis',
  'Jemuwah',
  'Setu',
  'Ngahad',
]

const WUKU = [
  'Sinta',
  'Landep',
  'Wukir',
  'Kurantil',
  'Tolu',
  'Gumbreg',
  'Warigalit',
  'Warigagung',
  'Julungwangi',
  'Sungsang',
  'Galungan',
  'Kuningan',
  'Langkir',
  'Mandasiya',
  'Julungpujut',
  'Pahang',
  'Kuruwelut',
  'Marakeh',
  'Tambir',
  'Medangkungan',
  'Maktal',
  'Wuye',
  'Manahil',
  'Prangbakat',
  'Bala',
  'Wugu',
  'Wayang',
  'Kulawu',
  'Dukut',
  'Watugunung',
]

const WINDU = ['Adi', 'Kuntara', 'Sengara', 'Sancaya']

const LAMBANG = ['Langkir', 'Kulawu']

const KURUP = [
  ['Jamingiyah', "A'ahgi", 1555, 0],
  ['Kamsiyah', 'Amiswon', 1675, 0],
  ['Arbangiyah', 'Aboge', 1749, 2],
  ['Salasiyah', 'Asapon', 1867, 0],
]

const ORIGIN = Date.UTC(1633, 6, 8)

const mod = (a, b) => ((a % b) + b) % b

const monthSum = (months) =>
  months.reduce((total, value) => total + value, 0)

function buildWT(maxYear = 2200) {
  const w354 = [30, 29, 30, 29, 30, 29, 30, 29, 30, 29, 30, 29]
  const w355 = [30, 29, 30, 29, 30, 29, 30, 29, 30, 29, 30, 30]

  const wtd1 = [30, 29, 30, 29, 30, 29, 30, 29, 30, 29, 30, 30]
  const wtd2 = [30, 30, 29, 29, 30, 29, 30, 29, 30, 29, 30, 30]
  const wtd3 = [30, 30, 29, 29, 29, 29, 30, 29, 30, 29, 30, 30]
  const wtd4 = [30, 29, 30, 29, 30, 29, 30, 29, 30, 29, 30, 29]

  const years = []

  for (let y = 1555; y <= maxYear; y += 1) {
    let k = KURUP.findIndex(
      (r, i) =>
        y >= r[2] &&
        (i === KURUP.length - 1 || y < KURUP[i + 1][2])
    )

    if (k < 0) k = KURUP.length - 1

    const pos = mod(y - 1555, 8)

    let months

    if (pos === 4) {
      months = [wtd1, wtd2, wtd3, wtd4][Math.min(k, 3)]
    } else if (pos === 1 || pos === 7) {
      months = w355
    } else if (pos === 3 && k >= 2) {
      months = w355
    } else {
      months = w354
    }

    years.push({ y, months: [...months], k })
  }

  for (let i = 0; i < KURUP.length; i += 1) {
    const end =
      i < KURUP.length - 1
        ? KURUP[i + 1][2] - 1
        : KURUP[i][2] + 119

    const record = years.find((year) => year.y === end)

    if (record) record.months[11] -= 1
  }

  return years
}

const YEARS = buildWT()

function jawaFromGregorian(y, m, d) {
  const dt = Date.UTC(y, m - 1, d)

  if (dt < ORIGIN) {
    throw new Error('Tanggal sebelum 8 Juli 1633 tidak didukung.')
  }

  const dti = Math.floor((dt - ORIGIN) / DAY_MS) + 1

  let remain = dti
  let record = null

  for (const year of YEARS) {
    const total = monthSum(year.months)

    if (remain <= total) {
      record = year
      break
    }

    remain -= total
  }

  if (!record) {
    throw new Error('Tanggal di luar rentang engine.')
  }

  let monthIndex = 0

  while (remain > record.months[monthIndex]) {
    remain -= record.months[monthIndex]
    monthIndex += 1
  }

  const day = remain
  const yearIndex = mod(record.y - 1555, 8)
  const serial = Math.floor(dt / DAY_MS)

  const utcWeekday = new Date(dt).getUTCDay()

  const dayName =
    MINGGU[(utcWeekday + 6) % 7]

  const pasaran =
    PAS[mod(serial + 2, 5)]

  const wuku =
    WUKU[
      mod(
        Math.floor((serial - 2) / 7) + 25,
        30
      )
    ]

  const kurupIndex = KURUP.findIndex(
    (r, i) =>
      record.y >= r[2] &&
      (i === KURUP.length - 1 || record.y < KURUP[i + 1][2])
  )

  const winduIndex =
    Math.floor((record.y - 1555) / 8)

  return {
    day,
    month: monthIndex + 1,
    monthIndex,
    monthName: WULAN[monthIndex],

    year: record.y,
    yearName: TAUN[yearIndex],

    dayName,
    pasaran,
    weton: `${dayName} ${pasaran}`,

    wuku,

    windu: WINDU[mod(winduIndex, 4)],
    lambang: LAMBANG[mod(winduIndex, 2)],

    kurup:
      KURUP[
        kurupIndex >= 0
          ? kurupIndex
          : KURUP.length - 1
      ][0],
  }
}

export function getJawaCalendar(context) {
  const date = context.time.instant

  const effectiveDate = new Date(
    Date.UTC(
      date.getFullYear(),
      date.getMonth(),
      date.getDate()
    )
  )

  return {
    ...jawaFromGregorian(
      effectiveDate.getUTCFullYear(),
      effectiveDate.getUTCMonth() + 1,
      effectiveDate.getUTCDate()
    ),

    effectiveDate,
    boundary: 'MIDNIGHT',

    meta: {
      engine: 'Jawa',
      phase: 'SOURCE_COMPILED',
      sunsetApplied: false,
    },
  }
}

File: backend/engines/jawa_engine.py
from datetime import date


WULAN = [
    "Sura", "Sapar", "Mulud", "Bakdamulud",
    "Jumadilawal", "Jumadilakir", "Rejeb", "Ruwah",
    "Pasa", "Sawal", "Dulkangidah", "Besar",
]

TAUN = [
    "Alip", "Ehe", "Jimawal", "Je",
    "Dal", "Be", "Wawu", "Jimakir",
]

PAS = ["Pon", "Wage", "Kliwon", "Legi", "Pahing"]

MINGGU = [
    "Senen", "Selasa", "Rebo", "Kemis",
    "Jemuwah", "Setu", "Ngahad",
]

WUKU = [
    "Sinta", "Landep", "Wukir", "Kurantil", "Tolu",
    "Gumbreg", "Warigalit", "Warigagung", "Julungwangi",
    "Sungsang", "Galungan", "Kuningan", "Langkir", "Mandasiya",
    "Julungpujut", "Pahang", "Kuruwelut", "Marakeh", "Tambir",
    "Medangkungan", "Maktal", "Wuye", "Manahil", "Prangbakat",
    "Bala", "Wugu", "Wayang", "Kulawu", "Dukut", "Watugunung",
]

WINDU = ["Adi", "Kuntara", "Sengara", "Sancaya"]
LAMBANG = ["Langkir", "Kulawu"]

KURUP = [
    ["Jamingiyah", "A'ahgi", 1555, 0],
    ["Kamsiyah", "Amiswon", 1675, 0],
    ["Arbangiyah", "Aboge", 1749, 2],
    ["Salasiyah", "Asapon", 1867, 0],
]

ORIGIN = date(1633, 7, 8)


def mod(a, b):
    return ((a % b) + b) % b


def month_sum(months):
    return sum(months)


def build_wt(max_year=2200):
    w354 = [30, 29, 30, 29, 30, 29, 30, 29, 30, 29, 30, 29]
    w355 = [30, 29, 30, 29, 30, 29, 30, 29, 30, 29, 30, 30]
    wtd1 = [30, 29, 30, 29, 30, 29, 30, 29, 30, 29, 30, 30]
    wtd2 = [30, 30, 29, 29, 30, 29, 30, 29, 30, 29, 30, 30]
    wtd3 = [30, 30, 29, 29, 29, 29, 30, 29, 30, 29, 30, 30]
    wtd4 = [30, 29, 30, 29, 30, 29, 30, 29, 30, 29, 30, 29]

    years = []

    for y in range(1555, max_year + 1):
        k = -1

        for i, row in enumerate(KURUP):
            if y >= row[2] and (
                i == len(KURUP) - 1 or y < KURUP[i + 1][2]
            ):
                k = i
                break

        if k < 0:
            k = len(KURUP) - 1

        pos = mod(y - 1555, 8)

        if pos == 4:
            months = [wtd1, wtd2, wtd3, wtd4][min(k, 3)]
        elif pos in (1, 7):
            months = w355
        elif pos == 3 and k >= 2:
            months = w355
        else:
            months = w354

        years.append({
            "y": y,
            "months": list(months),
            "k": k,
        })

    for i in range(len(KURUP)):
        end = (
            KURUP[i + 1][2] - 1
            if i < len(KURUP) - 1
            else KURUP[i][2] + 119
        )

        record = next(
            (year for year in years if year["y"] == end),
            None,
        )

        if record:
            record["months"][11] -= 1

    return years


YEARS = build_wt()


def jawa_from_gregorian(y, m, d):
    target = date(y, m, d)

    if target < ORIGIN:
        raise ValueError(
            "Tanggal sebelum 8 Juli 1633 tidak didukung."
        )

    dti = (target - ORIGIN).days + 1
    remain = dti
    record = None

    for year in YEARS:
        total = month_sum(year["months"])

        if remain <= total:
            record = year
            break

        remain -= total

    if record is None:
        raise ValueError("Tanggal di luar rentang engine.")

    month_index = 0

    while remain > record["months"][month_index]:
        remain -= record["months"][month_index]
        month_index += 1

    day = remain
    year_index = mod(record["y"] - 1555, 8)

    serial = (target - date(1970, 1, 1)).days
    utc_weekday = (target.weekday() + 1) % 7

    day_name = MINGGU[utc_weekday - 1] if utc_weekday > 0 else MINGGU[6]

    pasaran = PAS[mod(serial + 1, 5)]

    pawukon_day = mod(
        (target - date(2000, 1, 1)).days + 68,
        210,
    ) + 1

    wuku = WUKU[(pawukon_day - 1) // 7]

    kurup_index = -1

    for i, row in enumerate(KURUP):
        if record["y"] >= row[2] and (
            i == len(KURUP) - 1 or record["y"] < KURUP[i + 1][2]
        ):
            kurup_index = i
            break

    windu_index = mod(((record["y"] - 1555) // 8) + 1, 4)

    return {
        "day": day,
        "month": month_index + 1,
        "monthName": WULAN[month_index],
        "year": record["y"],
        "yearName": TAUN[year_index],
        "dayName": day_name,
        "pasaran": pasaran,
        "weton": f"{day_name} {pasaran}",
        "wuku": wuku,
        "windu": WINDU[windu_index],
        "lambang": LAMBANG[mod(windu_index, 2)],
        "kurup": KURUP[
            kurup_index if kurup_index >= 0 else len(KURUP) - 1
        ][0],
        "pawukonDay": pawukon_day,
    }


def get_jawa_data(target_date):
    result = jawa_from_gregorian(
        target_date.year,
        target_date.month,
        target_date.day,
    )

    return {
        **result,
        "effectiveDate": target_date.isoformat(),
        "meta": {
            "engine": "Jawa",
            "phase": "SOURCE_COMPILED",
        },
    }
